llvm/lib/Target/RISCV/RISCVAsmPrinter.cpp

```diff
@@ -363,13 +370,23 @@ bool RISCVAsmPrinter::emitDirectiveOptionArch() {
   return false;
 }
 
+bool RISCVAsmPrinter::doInitialization(Module &M) {
+
+  CommonSTI->setFeatureBits(TM.getMCSubtargetInfo()->getFeatureBits());
+  return AsmPrinter::doInitialization(M);
+}
+
 bool RISCVAsmPrinter::runOnMachineFunction(MachineFunction &MF) {
   STI = &MF.getSubtarget<RISCVSubtarget>();
   RISCVTargetStreamer &RTS =
       static_cast<RISCVTargetStreamer &>(*OutStreamer->getTargetStreamer());
 
   bool EmittedOptionArch = emitDirectiveOptionArch();
 
+  // Collect flags from this function.
+  CommonSTI->setFeatureBits(CommonSTI->getFeatureBits() |
```

Comment:
I don't think we can just union the features. Some features are incompatible, like F and 
Zfinx.

Reply:
> I don't think we can just union the features. Some features are incompatible, like F and Zfinx.

Hi @topperc, you have an interesting point. I tried a change in my test case, mixing incompatible feature flags, like:

```
attributes #0 = { nounwind "target-features"="+32bit,+d,+zicsr" }
attributes #1 = { nounwind "target-features"="+32bit,+d,+f,+m" }
attributes #2 = { nounwind "target-features"="+32bit,+f,+c" }
attributes #3 = { nounwind "target-features"="+32bit,+a,+zfinx" }
```
Without the PR, we generated the ` Value: rv32i2p1 ` in the elf, but with the PR, I got:

```
LLVM ERROR: 'f' and 'zfinx' extensions are incompatible

PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump: [...]
```
Do you think it is a useful case to have mixed (and even incompatible) flags in the same module? I only can think about some multicore asymmetric environments...

Reply:
`f` and `zfinx` are most obviously case in standard RISC-V extension, but may not common in real world, the most common case I could imagine is the multi-version function with different vendor extensions, and those vendor extension are using overlapped encoding.

One example is T-head Vector and standard vector extension (IIRC they didn't upstream to LLVM yet but they has planned).

```c
void foo_thead __attribute__((target="arch=+theadvector")) ()
{
}
void foo_v __attribute__((target="arch=+v")) ()
{
}

void bar(){
  if (thead_vector_ext_ok ())
    foo_thead ()
  if (std_vec_ext_ok ())
    foo_v ()
}
```

Reply:
Hi @kito-cheng, thank you for this point, makes sense. I will include in this PR a mechanism to exclude invalid feature bit combinations.